embassy-stm32/src/dma/dma.rs

```diff
@@ -273,9 +494,38 @@ mod low_level_api {
                 dma.0 as u32, channel_num
             );
         }
-        if isr.tcif(channel_num % 4) && cr.read().tcie() {
-            cr.write(|_| ()); // Disable channel interrupts with the default value.
-            STATE.ch_wakers[index].wake();
+
+        if GIANT_TRANSFER_STATE.is_enabled_for_channel(state_index) {
+            if isr.tcif(channel_num % 4) && cr.read().tcie() {
+                dma.ifcr(channel_num / 4)
+                    .write(|w| w.set_tcif(channel_num % 4, true));
+
+                let remaining_transfers =
+                    GIANT_TRANSFER_STATE.remaining_chunks_for_channel(state_index);
+                if remaining_transfers != 0 {
+                    let current_target_memory1 = cr.read().ct() == vals::Ct::MEMORY1;
+                    if remaining_transfers % 2 == 0 && current_target_memory1 {
+                        // update pointer to memory 0
+                        let mem0_addr = dma.st(channel_num).m0ar().read();
+                        let new_addr =
+                            GIANT_TRANSFER_STATE.dequeue_next_chunk(state_index, mem0_addr);
+                        dma.st(channel_num).m0ar().write_value(new_addr);
+                    } else if !current_target_memory1 {
+                        // update pointer to memory 1
+                        let mem1_addr = dma.st(channel_num).m1ar().read();
+                        let new_addr =
+                            GIANT_TRANSFER_STATE.dequeue_next_chunk(state_index, mem1_addr);
+                        dma.st(channel_num).m1ar().write_value(new_addr);
+                    }
+                } else {
+                    GIANT_TRANSFER_STATE.disable_for_channel(state_index);
```

Comment:
where is the DMA hardware actually stopped? AFAICT it'll keep reading...?

Reply:
Oh, is it the `cr.write(|_| ());`? I see. Would be nice to clarify it with a comment, the current comment indicates it's just to disable irqs.

Isn't it "too late" by then though? like, after you get the last TCIF, the hardware won't stop automatically, it'll swap buffers once again and start overwriting the second-to-last chunk.

Reply:
Good thinking. I had a look at the Cube code and it seems to set M0AR to the initial buffer address and M1AR to first chunk address https://github.com/STMicroelectronics/STM32CubeH7/blob/b340b13929e36a3427b8d94e8b1006022f82273f/Drivers/STM32H7xx_HAL_Driver/Src/stm32h7xx_hal_dcmi.c#L1151-L1161

Which is I assume kinda works for DCMI as it doesn't matter that much if the data get overwritten? Definitely feels weird.

Reply:
Yeah, seems like they ignore the issue there, probably because in the case of DCMI you control how much data it will generate, so if you read exactly 1 frame it won't actually overwrite stuff. This is not the case for UART etc.

Maybe the "poison address" Idea would work here? At second-to-last TCIF, set the other MxAR to 0xFFFF_FFFF, At the last TCIF, it'll switch to writing there, and stop due to the error.